compiler/qsc_rca/src/lib.rs

```diff
@@ -0,0 +1,841 @@
+                write!(f, "[Parameter Type Array] {array_param_application}")?;
+            }
+        };
+        Ok(())
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct ArrayParamApplication {
+    pub static_content_dynamic_size: ComputeKind,
+    pub dynamic_content_static_size: ComputeKind,
+    pub dynamic_content_dynamic_size: ComputeKind,
+}
+
+impl Display for ArrayParamApplication {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        let mut indent = set_indentation(indented(f), 0);
+        write!(indent, "ArrayParamApplication:",)?;
+        indent = set_indentation(indent, 1);
+        write!(
+            indent,
+            "\nstatic_content_dynamic_size: {}",
+            self.static_content_dynamic_size
+        )?;
+        write!(
+            indent,
+            "\ndynamic_content_static_size: {}",
+            self.dynamic_content_static_size
+        )?;
+        write!(
+            indent,
+            "\ndynamic_content_dynamic_size: {}",
+            self.dynamic_content_dynamic_size
+        )?;
+        Ok(())
+    }
+}
+
+#[derive(Clone, Copy, Debug)]
+pub enum ComputeKind {
+    Classical,
+    Quantum(QuantumProperties),
+}
+
+impl Display for ComputeKind {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        match &self {
+            ComputeKind::Quantum(quantum_properties) => write!(f, "Quantum: {quantum_properties}")?,
+            ComputeKind::Classical => write!(f, "Classical")?,
+        };
+        Ok(())
+    }
+}
+
+impl ComputeKind {
+    pub(crate) fn new_with_runtime_features(
+        runtime_features: RuntimeFeatureFlags,
+        value_kind: ValueKind,
+    ) -> Self {
+        Self::Quantum(QuantumProperties {
+            runtime_features,
+            value_kind,
+        })
+    }
+
+    pub(crate) fn aggregate(self, value: Self) -> Self {
+        let ComputeKind::Quantum(value_quantum_properties) = value else {
+            // A classical compute kind has nothing to aggregate so just return self with no changes.
+            return self;
+        };
+
+        // Determine the aggregated runtime features.
+        let runtime_features = match self {
+            Self::Classical => value_quantum_properties.runtime_features,
+            Self::Quantum(ref self_quantum_properties) => {
+                self_quantum_properties.runtime_features | value_quantum_properties.runtime_features
+            }
+        };
+
+        // Determine the aggregated value kind.
+        let value_kind = match self {
+            Self::Classical => value_quantum_properties.value_kind,
+            Self::Quantum(self_quantum_properties) => self_quantum_properties
+                .value_kind
+                .aggregate(value_quantum_properties.value_kind),
+        };
+
+        // Return the aggregated compute kind.
+        ComputeKind::Quantum(QuantumProperties {
+            runtime_features,
+            value_kind,
+        })
+    }
+
+    pub(crate) fn aggregate_runtime_features(
```

Comment:
I would've expected this to take or produce `RuntimeFeatureFlags` since that's what other functions mean when they talk about "runtime_features" but these are instead calculated internally. Not sure how to convey this in the name, but really it calculates and then aggregates the runtime features.

Reply:
Yes, this one is tricky. I also added it to #1285 to reconsider what name should we give to this function.